M3-207 - Checkbox Component UI Test

patthiel opened the pull request:
* Added storybook webdriverIO configuration file (should be separate config from the standard wdio manager config since different spec folder paths, baseUrls, pre-test setup hooks, etc). Inherit everything from wdio.conf.js and override specific config values.
* Added jsClick and jsClickAll browser custom commands, useful when attempting to click an element throws a selenium error like "other element would receive click"
* Bumped chromedriver version to latest.
* Added storybook homepage route to constants.js
* Created test suite for Checkbox UI Component
* Added `data-qa` attribute to checkbox component
* Added `storybook:e2e` yarn script.
* Updated docker scripts to remove containers after they exit, and to no longer hardcode the container names.

**NOTE:**  i made the component ui test specs live inside each component folder. I think this sort of test (especially since they can be run very quickly) should live longside the component in the code base. Since it is not typescript though, it will not be linted via `yarn lint`. I can configure eslint in a branch to lint e2e tests contained in here though

## To Test:
1. Start storybook: `yarn docker:storybook`
2. (New shell) Start selenium: `yarn selenium`
3. (New shell) Run the tests: `yarn storybook:e2e`

patthiel commented on src/components/CheckBox/CheckBox.spec.js:
```diff
@@ -0,0 +1,71 @@
+    it('should check on click', () => {
+        browser.jsClickAll('[data-qa-checked] input');
+
+        previewFocus();
```
In the coming PR's i have changed this to no longer call `previewFocus` multiple times in tests, but instead i've created a method that will call `browser.frame` and `browser.isExisting(selector)` for a fixed period of time until `browser.isExisting(selector)` returns true.

asauber commented on src/components/CheckBox/CheckBox.spec.js:
```diff
@@ -0,0 +1,71 @@
+        const updatedCheckboxValues = $$('[data-qa-checked]');
+        const boxValues = updatedCheckboxValues.map(e => e.getAttribute('data-qa-checked') == 'true');
+        
+        expect(boxValues).toContain(true);
```
This doesn't test that all of the checkboxes have been checked, only one or more of them

patthiel replied:
Correct. I can't guarantee that it will check x number of boxes. So if the story file is modified and 5 are checkable now, this test would break. I wanted to test that clicking all the check boxes results in at least one checkbox checking. 

Thinking about this some more, I could create an array of check box elements that do not contain the "disabled" class and click those. That way I would have a true expected number of boxes that can be checked. What do you think @asauber ?

asauber replied:
Right, ok. Rather than have tests be based on counting, we can use reduce to test if all of the values are true or false. This will depend on selecting only the non-disabled checkboxes to start.

asauber commented on src/components/CheckBox/CheckBox.spec.js:
```diff
@@ -0,0 +1,71 @@
+        previewFocus();
+        previewFocus();
+
+        const boxes = $$('[data-qa-checked]');
```
The name `data-qa-checked` would make more sense as `data-qa-checkbox`

patthiel replied:
agreed, thanks

asauber commented on src/components/CheckBox/CheckBox.spec.js:
```diff
@@ -0,0 +1,71 @@
+
+        const boxes = $$('[data-qa-checked]');
+        const numberDisabledBoxes = boxes.map(e => e.getAttribute('class').includes('disabled'));
+        expect(numberDisabledBoxes).toContain(true);
```
I would expect this test to count the number of disabled checkboxes

patthiel replied:
This again speaks to how specific we should or shouldn't make these tests. Do we want to always expect x number of boxes to be disabled in this test, or just _some_ boxes are disabled. Up to you guys, I can certainly modify the test to be more specific.

asauber commented:
Looks great, tests are easy to run. A few requested changes.

rjstires commented:
@patthiel is this still in progress?

patthiel commented:
@rjstires i made the changes @asauber requested and changed how to the check all enabled checkboxes test works. 

I also added tests attempting to check disabled checkboxes per @asauber feedback as well. Otherwise this is good to go

rjstires commented:
Too @asauber for the final 👍then.
